docs/dg/dev/guidelines/performance-guidelines/database-performance-quidelines.md

```diff
@@ -0,0 +1,133 @@
+    {
+        if ($this->statusChanged && $this->getIdSalesOrderItem()) {
+            /** @var \Orm\Zed\Sales\Persistence\SpySalesOrderItem $salesOrderItemEntity */
+            $salesOrderItemEntity = $this;
+            $omsOrderItemStateHistoryEntity = $this->createOmsOrderItemStateHistoryEntity();
+            $omsOrderItemStateHistoryEntity->setOrderItem($salesOrderItemEntity);
+            $omsOrderItemStateHistoryEntity->setFkOmsOrderItemState($this->getFkOmsOrderItemState());
+
+            $this->sharedPersist($omsOrderItemStateHistoryEntity);
+        }
+
+        $this->statusChanged = false;
+    }
+}
+```
+With this implementation, the following code:
+```php 
+public function updateOrCreateSalesOrderItems(array $salesOrderItemsData): vois
+{
+    // Use batch processor trait for efficient processing
+    foreach ($salesOrderItemsData as $orderItemData) {
+        // Find or create the sales order item
+        $salesOrderItem = $this->findOrCreateSalesOrderItem($orderItemData);
+
+        // Update item attributes
+        $this->updateSalesOrderItemAttributes($salesOrderItem, $orderItemData);
+
+        // Collect for batch processing
+        $this->persist($salesOrderItem);
+    }
+
+    // Commit all collected items in a single batch operation
+    $this->commit();
+}
+```
+
+Performs the following actions:
+- Loads all required entities from the database in a single query.
+- Creates or updates `salesOrderItem` entities in a single batch. 
+- Creates related `OrderItemStateHistoryEntity` instances during the post-save process and saves them in a single operation after the main entity save is complete.
+
+### BatchEntityHooksInterface
+- Enables `postSave` hooks similar to publish-and-synchronize (P&S) events.
+- Should be implemented for storage and search entities to ensure proper event handling during batch operations.
+
+#### Basic Usage Example:
+
+```php
+class AbstractSpyProductOfferStorage extends BaseSpyProductOfferStorage implements BatchEntityHooksInterface
+{
+    public function batchPreSaveHook(): void
+    {
+        if (method_exists($this, 'isSynchronizationEnabled') && $this->isSynchronizationEnabled()) {
+            // synchronization behavior
+            $this->setGeneratedKey();
+            $this->setGeneratedKeyForMappingResource();
+            $this->setGeneratedAliasKeys();
+        }
+    }
+```
+When saving the `SpyProductOfferStorage` entity using `ActiveRecordBatchProcessorTrait`, the P&S event is triggered after the batch save completes.
+
+#### Limitations and Suggestions:
+Limitations and Recommendations
+1. **Entity ID Access**: The `ActiveRecordBatchProcessorTrait` does not return entity IDs after saving. If you need the ID, perform a separate database query.
+
+2. **P&S Event Compatibility**: `ActiveRecordBatchProcessorTrait` does not trigger publish-and-synchronize events for storage or search entities. Always implement `BatchEntityHooksInterface` for each such entity.
```

Comment:
As mentioned in a comment above. That doesn't sound neither like limitation not suggestion. It is a requirement, which is part of "how this feature works"